### PR TITLE
Optimize rowset usage

### PR DESCRIPTION
**Pull request by kangpinghuang**
1. optimize incremental rowsetshared ptr ref to aquire the rowset object
2. add unused_rowset_thread to delete unused rowset
3. add need_delete_flag to mark deleted rowset
4. optimize AlphaRowsetReader's constructor

### CODE REVIEW

File: be/src/olap/rowset/rowset.h [hunk 1]
```diff
@@ -113,13 +113,23 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
...
     virtual bool check_path(const std::string& path) = 0;
 
+    bool need_delete_file() {
+        return _need_delete_file;
+    }
+
+    void set_need_delete_file(bool need_delete_file) {
+        _need_delete_file = need_delete_file;
+    }
+
```
Comment (yiguolei) [inline be/src/olap/rowset/rowset.h]: add a check here. if it is set need deleted it could not be reverted.

File: be/src/olap/storage_engine.cpp [hunk 2]
```diff
@@ -677,17 +675,19 @@ OLAPStatus StorageEngine::_do_sweep(
 
 void StorageEngine::start_delete_unused_rowset() {
     _gc_mutex.lock();
-
-    auto it = _unused_rowsets.begin();
-    for (; it != _unused_rowsets.end();) { 
-        if (it->second->in_use()) {
+    LOG(INFO) << "start to delete unused rowset.";
```
Comment (chaoyli) [inline be/src/olap/storage_engine.cpp]: Log is non-sense when has no unused rowset.

File: be/src/olap/storage_engine.cpp [hunk 3]
```diff
@@ -677,17 +675,19 @@ OLAPStatus StorageEngine::_do_sweep(
...
+    for (auto it = _unused_rowsets.begin(); it != _unused_rowsets.end();) {
+        if (it->second.use_count() != 1) {
             ++it;
-        } else {
-            it->second->remove();
-            _unused_rowsets.erase(it);
+        } else if (it->second->need_delete_file()){
+            LOG(INFO) << "start to remove rowset:" << it->second->rowset_id()
+                    << ", version:" << it->second->version().first << "-" << it->second->version().second;
+            OLAPStatus status = it->second->remove();
```
Comment (chaoyli) [inline be/src/olap/storage_engine.cpp]: if failed, you can not remove it from unused rowset.

File: be/src/olap/rowset/rowset.h [hunk 4]
```diff
@@ -113,13 +113,26 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
 
     virtual int64_t txn_id() const = 0;
     
+    // flag for push delete rowset
     virtual bool delete_flag() = 0;
 
     virtual bool check_path(const std::string& path) = 0;
 
+    bool need_delete_file() {
+        return _need_delete_file;
```
Comment (chaoyli) [inline be/src/olap/rowset/rowset.h]: if _need_delete_file = false and rowset shared_ptr use_count() == 1 will happen meanwhile?

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Pull request merged**